### ISSUE TITLE
`jj resolve` should handle rename+edit conflicts nicely

### ISSUE THREAD

**Issue by dbarnett** [martinvonz/jj#1112]
## Description

When merging one revision that moves (or copies) a file and another that makes changes in the original file, the `jj resolve` command should offer more help in diffing the right files and getting all the changes integrated where they belong.

This also entails jj tracking some notion of tracking file move/copies (#47). Although the git backend might have some rename detection of its own?

### Trivial scenario

* Rev 1 just renames a.txt to b.txt
* Rev 2 edits a.txt
* Merge commit 1+2 should contain b.txt with the contents of rev 2's a.txt (ideally handled fully automatically)

### Rename-edit + edit scenario

* Rev 1 renames a.txt to b.txt *and makes edits*
* Rev 2 edits a.txt
* Merge commit 1+2 should contain b.txt and prompt to resolve conflicts, showing a diff between rev 1's b.txt and rev 2's *a.txt*

### Copy-edit + edit scenario

* Rev 1 copies a.txt to b.txt and makes edits in both
* Rev 2 edits a.txt
* Merge commit 1+2 should contain both a.txt and b.txt and prompt to resolve conflicts between *both* of the following pairs (because rev 2's edits might ultimately belong in a.txt or b.txt, or both):
  * rev 1's a.txt and rev 2's a.txt
  * rev 1's *b.txt* and rev 2's a.txt

In some cases it might already be unambiguous whether rev 2's edits belong in a.txt or b.txt and not need to resolve against both. For instance, if rev 1 had forked the contents of a.txt into 2 files, keeping lines 1-50 in a.txt and moving lines 51-100 into b.txt, then it might be self-evident how to properly incorporate other edits into one destination file or the other.

**Comment by sageserpent-open** [comment]
@dbarnett **Kinetic Merge** handles this, including the situation where `a.txt` is split into two files, one half of the content remaining at the same path and the other going into a new file.

I don’t know my way around JJ at all (just reading the issue tickets to get a feel for the state of the art), but I gather that one can interoperate between JJ and Git in the same repository or worktree or whatever.

If that is the case, you could run Kinetic Merge (because that works with Git repositories) and switch back to JJ.

Not what you’re asking for in this ticket, but possibly a solution to the problems you’re describing.

**Comment by lypanov** [comment]
At the very least some mention of how to move ahead with fixing it / digging into the commits involved etc. Running into this now and jj is certainly not 'auto documenting' in this specific conflict situation even though elsewhere it tries it's best. Far from in fact, I simply have no idea what to do and it's been 20 minutes since I started on this.

**Comment by sageserpent-open** [comment]
@lypanov As stated above, I'm not conversant with jj, but from my limited understanding it should be possible to either make a local clone of your repository, or at very least to push changes to a remote repository and then clone from *that* repository into a local clone to merge there via Git + Kinetic Merge or whatever.

The idea being to set up the local clone as being a colocated one. I gather that turning a non-colocated repository into a colocated is considered possible but is not officially supported, but I thought colocation from the start *is* supported?

(https://github.com/jj-vcs/jj/blob/main/docs/git-compatibility.md)

I'm not suggesting this is a desirable workflow at all, but it is potentially a workaround for others who occasionally find themselves with a tricky merge.

There is a ticket over in Kinetic Merge about jj integration, but that is on hold right now as the requirements are up in the air: https://github.com/sageserpent-open/kineticMerge/issues/136.

**Comment by Kinrany** [comment]
From https://github.com/jj-vcs/jj/issues/6940:

> In practice what I'd like to be able to do when I see a conflict that includes a deletion, is to manually apply the conflicts to the file at the new location. So that the updated file at the new location will contain all the conflicts but without the deletion.
>
> (If moves can be detected automatically and when there's only one location, I think this could be done automatically. The only decision that needs to be made is the choice of the location to keep.)
>
> More generally, since conflicts are first-class, it would be good if the CLI supported operations on files with conflicts in them.

**Comment by Kinrany** [comment]
One more thing that is missing is the ability to move a file with conflicts in it. Currently jj ignores the conflict markers in the file at the new location, which is very mysterious. I still have no idea if it's possible to tell jj explicitly to treat the conflict markers in the new file as real conflicts.

**Comment by martinvonz** [comment]
Good point. I don't think there's a way of doing that yet. Once we have a `jj file rename` command, I think that will handle this case.
